datafusion/sql/src/expr/mod.rs

```diff
@@ -168,6 +169,7 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
 
     /// Internal implementation. Use
     /// [`Self::sql_expr_to_logical_expr`] to plan exprs.
+    #[recursive]
```

Comment:
you should be able to remove the note in `sql_array_literal` about stack overflow